[browser][wasm] JavaScript core Map object disposal - Do Not Merge

kjpou1 opened the pull request:
This fixes a crash that was happening when garbage collection kicked in.  When allocating larger key value pair counts the Map object would throw a NRE intermittently.  

Also, added larger iteration tests to the map tests so this can be caught in the future.

There seems to be another problem somewhere and needs to be looked at further.

kjpou1 committed:
Remove finalizer code from Map object.

kjpou1 committed:
Add larger iteration tests to make sure we check for GC kicking in.

campersau commented:
Was that the cause of https://github.com/dotnet/runtime/pull/39061 so `MapTests.cs` can be enabled again?
https://github.com/dotnet/runtime/blob/9745dd7e7dba725faed70a0258f5965557398a0f/src/libraries/System.Runtime.InteropServices.JavaScript/tests/System.Runtime.InteropServices.JavaScript.Tests.csproj#L15

kjpou1 committed:
Enable the map tests again

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-map-disposal

# Conflicts:
#	src/libraries/System.Runtime.InteropServices.JavaScript/tests/System.Runtime.InteropServices.JavaScript.Tests.csproj

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-map-disposal

kjpou1 commented:
@marek-safar could you take a look ?

marek-safar commented on src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Map.cs:
```diff
@@ -147,16 +147,9 @@ private void Dispose(bool disposing)
                 }
             }
 
-            ~MapEnumerator()
```
I don't understand how can this fix the problem. It only disables the tests coverage,

kjpou1 replied:
The tests were failing with null object when GC kicked in.  The taking this out it the null object error went away when GC kicked in.

lewing reviewed (commented):
The dispose logic doesn't make much sense now, this is a sealed class with only a public Dispose method and an indirect reference to an an unmanaged object that is guarded against reset in a strange way.  Can you explain why removing the finalizer is the correct fix?

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-map-disposal

kjpou1 committed:
Cleanup disposing code.

- The internal classes are sealed so with out the ~ finalizer the variable checking the variable `disposing` is no longer necessary.

kjpou1 commented:
Worked on the disposing logic but there seems to be another error. 

```
JS exception: Error: no idea on how to unbox value types
        
  dbug: test[0]
        Error: no idea on how to unbox value types
        
  dbug: test[0]
            at Object.unbox_mono_obj (dotnet.js:7840:13)
```

lewing commented:
@kjpou1 if this is active please rebase it otherwise please close the pr

kjpou1 commented:
No longer valid as other PR resolved issues and map tests are enabled again.
